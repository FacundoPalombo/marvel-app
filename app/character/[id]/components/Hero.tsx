"use client";

import { FavoritesContext } from "@/app/favorites-context";
import styles from "./Hero.module.css";

import { HeartFilled, HeartOutline } from "@/app/components/svg/Heart";
import { useCallback, useEffect, useState } from "react";

type HeroProps = {
  img: string;
  name: string;
  description: string;
  id: string;
};
export default function Hero({ img, description, name, id }: HeroProps) {
  const [favorites, setFavorites] = useState<string[]>(() => {
    const favs = window.localStorage.getItem("favorites");
    return JSON.parse(favs as string) || [];
  });

  useEffect(() => {
    window.localStorage.setItem("favorites", JSON.stringify(favorites));
  }, [favorites]);

  const handleFavorite = useCallback(() => {
    if (favorites.includes(id)) {
      setFavorites(favorites.filter((fav) => fav !== id));
    }
    if (!favorites.includes(id)) {
      setFavorites([...favorites, id]);
    }
  }, [favorites]);

  return (
    <FavoritesContext.Provider value={{ favorites, setFavorites }}>
      <section className={styles.hero}>
        <div className={styles.hero__container}>
          <img className={styles.hero__image} src={img} alt={name} />
          <article className={styles.hero__content}>
            <div className={styles.hero__heading}>
              <h1 className={styles.hero__title}>{name}</h1>
              <button
                className={styles.hero__favorite}
                onClick={handleFavorite}
              >
                {favorites.includes(id) ? (
                  <HeartFilled />
                ) : (
                  <HeartOutline className={styles.hero__heartOutline} />
                )}
              </button>
            </div>
            <p className={styles.hero__description}>{description}</p>
          </article>
        </div>
        <div className={styles.hero__borderShape} />
      </section>
    </FavoritesContext.Provider>
  );
}
