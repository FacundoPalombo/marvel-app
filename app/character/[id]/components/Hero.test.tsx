import { render } from "@testing-library/react";
import Hero from "./Hero";

describe("Hero unit tests", () => {
  it("should match to snapshot", () => {
    const props = {
      img: "image",
      name: "Tremendo Hero",
      issueDate: "2024",
      description: "lorem ipsum",
      favorite: true,
    };

    const { asFragment } = render(<Hero {...props} />);

    expect(asFragment()).toMatchSnapshot();
  });
  it("should render ok", () => {
    const props = {
      img: "image",
      name: "Tremendo Hero",
      issueDate: "2024",
      description: "lorem ipsum",
      favorite: true,
    };

    const { getByRole } = render(<Hero {...props} />);

    expect(getByRole("heading", { name: "Tremendo Hero" })).toBeInTheDocument();
  });
});
