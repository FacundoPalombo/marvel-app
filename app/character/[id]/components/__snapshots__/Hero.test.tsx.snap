// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hero unit tests should match to snapshot 1`] = `
<DocumentFragment>
  <section
    class="hero"
  >
    <div
      class="hero__container"
    >
      <img
        alt="Tremendo Hero"
        class="hero__image"
        src="image"
      />
      <article
        class="hero__content"
      >
        <div
          class="hero__heading"
        >
          <h1
            class="hero__title"
          >
            Tremendo Hero
          </h1>
          <button
            class="hero__favorite"
          >
            <svg
              class=""
              fill="none"
              height="22"
              viewBox="0 0 24 22"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12 3.63869L6 -0.00292969L0 3.63869V11.4422L12 21.6734L24 11.4422V3.63869L18 -0.00292969L12 3.63869Z"
                fill="#EC1D24"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
        <p
          class="hero__description"
        >
          lorem ipsum
        </p>
      </article>
    </div>
    <div
      class="hero__borderShape"
    />
  </section>
</DocumentFragment>
`;
